File: codes/naeun/84_largestrectangleinhistogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int result = heights[0]; // width * min height
        Stack<Integer> stack = new Stack<>();
        stack.push(0);

        for(int i = 1; i < heights.length; i++){
            while(!stack.isEmpty() && heights[stack.peek()] <= heights[i]){
                int prevIdx = stack.pop();
                result = Math.max((i-prevIdx+1) * heights[prevIdx], result);
            }
            result = Math.max(result, heights[i]);
            stack.push(i);
        }

        return result;
    }
}

File: codes/jiheon/84_largestrectangleinhistogram.rs
impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut stack: Vec<usize> = Vec::new();
        let mut ans: i32 = 0;
        let n = heights.len();
        for (i, current_height) in heights.iter().enumerate() {
            while matches!(stack.last(), Some(ele) if heights[*ele] > *current_height) {
                let prev_largest_idx = stack.pop().unwrap();
                let height = heights[prev_largest_idx];
                let width = if stack.is_empty() {
                    i
                } else {
                    i - stack.last().unwrap() - 1
                } as i32;
                ans = ans.max(width * height);
            }
            stack.push(i);
        }

        while let Some(prev_largest_idx) = stack.pop() {
            let height = heights[prev_largest_idx];
            let width = if stack.is_empty() {
                n
            } else {
                n - stack.last().unwrap() - 1
            } as i32;
            ans = ans.max(width * height);
        }

        ans
    }
}